File: liverepl-server/src/net/djpowell/liverepl/server/Repl.java
package net.djpowell.liverepl.server;

import java.net.InetAddress;
import java.lang.instrument.Instrumentation;

import clojure.lang.Namespace;
import clojure.lang.PersistentArrayMap;
import clojure.lang.RT;
import clojure.lang.Symbol;
import clojure.lang.Var;

public class Repl {
    private final static Symbol REPL_NS = Symbol.create("net.djpowell.liverepl.server.repl");
    private final static Namespace NS = Namespace.findOrCreate(REPL_NS);
    private final static Var REQUIRE = Var.intern(RT.CLOJURE_NS, Symbol.create("require"));
    private final static Var REPL = Var.intern(NS, Symbol.create("repl"));

    private final static Symbol AGENT_NS_NAME = Symbol.create("jvm.agent");
    private final static Namespace AGENT_NS = Namespace.findOrCreate(AGENT_NS_NAME);
    
    public static void main(InetAddress host, int port, Instrumentation inst)
	throws Exception {
	final Var instv = Var.intern(AGENT_NS, Symbol.create("instrumentation"));
	instv.bindRoot(inst);
	REQUIRE.invoke(REPL_NS);
	REPL.invoke(port, 0, host);
    }
}

File: liverepl-server/src/net/djpowell/liverepl/server/repl.clj
(ns net.djpowell.liverepl.server.repl
  (:require clojure.main)
  (:import (java.io InputStreamReader BufferedWriter OutputStreamWriter Writer)
    (java.net ServerSocket Socket)
    (java.util.logging Logger)
    (clojure.lang LineNumberingPushbackReader)))

(def new-line (System/getProperty "line.separator"))

(defn repl-caught
  "Override the default repl-caught to not rely on PrintStream"
  [e]
  (.write #^Writer *err* (str (clojure.main/repl-exception e) new-line)))

;; The client should connect immediately after the server starts up,
;; but in case there is a problem, we will kill the serversocket,
;; terminating the server, if the client fails to connect.
(def client-connect-timeout 10000)
  
(def #^Logger trc (Logger/getLogger (str *ns*)))
  
(defn repl
  "Start a single connection repl server"
  [port backlog host]
  (let [server (ServerSocket. port backlog host)
    got-connection (atom false)
    killer-thread (Thread.
      (fn []
        (Thread/sleep client-connect-timeout)
          (when-not @got-connection
            (.fine trc "Client connect timeout: terminating server")
            (.close server)))
      "Killer Thread")
    repl-thread (Thread.
      (fn []
        (with-open [server server
          socket (.accept server)
          in (LineNumberingPushbackReader. (InputStreamReader. (.getInputStream socket)))
          out (BufferedWriter. (OutputStreamWriter. (.getOutputStream socket)))]
            (swap! got-connection (constantly true))
            (binding [*in* in
              *out* out
              *err* out]
                (try
                  (clojure.main/repl :caught repl-caught :init #(in-ns 'user))
                (catch Exception e ; discard errors caused by abrupt disconnection
                  nil)))))
      "Repl Thread")]
    (.start killer-thread)
    (.start repl-thread)))
